SPICE-0029: Self Type (#32)
apple/pkl-evolution: spices/SPICE-0029-self-type.adoc

== Detailed design

Pkl's parsers will be updated to recognize `this` in type positions.
Similarly, pkl-intellij and pkl-lsp (incl. tree-sitter-pkl) will be updated to parse and resolve `this` as a type.

=== Prior Art & Naming

Several other languages provide a self type and write it in different ways:

* Swift: `Self`
* Rust: `Self`
* TypeScript: `this`
* Python: `Self`
* Zig: `@This()` (often aliased to `Self` via `const Self = @This()`)

Swift, Rust, and Python have a `self`/`Self` distinction: `self` talks about the instance, and `Self` talks about the type.

This SPICE opts to write Pkl's self type as `this` for a few reasons:

* Its casing aligns with Pkl's existing keyword types (`module`, `unknown`, `nothing`).
* Using `this` avoids allocating a new keyword, which is a breaking change for code using that name as an identifier.
** Choosing `self` is particularly problematic as it's common in "scope capture" scenarios often written as `local self = this`.
* Using `this` to mean a type or a value depending on context matches the existing behavior for `module`.

=== Semantics

The `this` type is an alias for the class of the current receiver (`this` value).

> For open classes, `this` admits values whose class is a subclass of the receiver.

Reviewer: ```suggestion
For open or abstract classes, `this` is the specific subclass of that class
```

= Self Type

* Proposal: link:./SPICE-0029-self-type.adoc[SPICE-0029]
* Author: https://github.com/HT154[Jen Basch]
* Status: TBD
* Implemented in: Pkl 0.33
* Category: Language, Tooling

== Introduction

This SPICE proposes a new, general-use self type.

== Motivation

Pkl currently has a self type, `module`, that is only applicable in specific situations.
As a result, it's currently possible to express patterns using modules that cannot be expressed by normal classes.

The `module` type also means different things at the module level vs. inside a class or typealias body, which can be confusing.

== Proposed Solution

A new type `this` will be added that serves as a self type in both module and class contexts.

== Detailed design

Pkl's parsers will be updated to recognize `this` in type positions.

> Similarly, pkl-intellij and pkl-lsp (incl. tree-sitter-pkl) will be updated to parse and resolve `this` as a type.

Reviewer: Can we include sections around:

* Prior art and naming

  Swift, Rust, TypeScript, Python, and Zig each have the concept of "self type":

  Swift: `Self`
  Rust: `Self`
  TypeScript: `this`
  Python: `Self`
  Zig: `const Self = @This()`

  Swift/Rust/Python have a `self/Self` distinction: `self` talks about the instance, and `Self` talks about the type.

  However, `this` feels more appropriate for Pkl for both the type and the instance because we already have this parity with `module` (it talks about both the instance and the type). Also, this is nice because it's a non-breaking change.

* Type checking semantics (what is well typed, what becomes type errors, etc)

Author: I think I covered the semantics in enough depth, but let me know if there's anything I need to spend more detail on!

=== Usage Restrictions

Semantics of self types inside type alias bodies are not easily definable, especially when the enclosing module is not final.

> Usage of both `this` in such contexts violates the existing restriction on accessing only `const` values.

Reviewer: I think mentioning `const` here is a little confusing. Maybe:

```suggestion
A self type is not compatible with typealiases.
A typealias is meant be "static"; an extending module does not change the meaning of a typealias (see https://github.com/apple/pkl-evolution/blob/main/spices/SPICE-0007-const-checks-in-typealiases.adoc[SPICE-0007]).
On the other hand, the `this` type always refers to the type of the extending module.
Because of this, the `this` type cannot be used inside typealias bodies.
```

As such, the `this` type is forbidden for use inside typealias bodies.

The `module` type has similar issues when used inside class bodies and typealias.
A new deprecation warning for use of the `module` type in these contexts will be introduced and will become an evaluation error in a subsequent release.

> In most scenarios, the `module` type should be replaced with a self-import.

Reviewer: Some thoughts:

* Can you add something to the intro to the line of "this SPICE also refines the module type"?
* I think this needs its own section; what does "deprecation warning" mean? What are examples of invalid usage of the `module` type?
* How long is the deprecation cycle?
* Should we provide a migrator? Quickfix?

New diagnostics will be added to pkl-intellij and pkl-lsp highlighting these usages

=== Pkl API

Only a couple Pkl API changes are directly required as part of adding the `this` type:

* Add `pkl.reflect#ThisType`
* Add `pkl.reflect#thisType`

A couple more changes will be made to adopt the `this` type in existing APIs:

* `Any.getClass()` will return `Class<this>` instead of just `Class`
** This allows removing special-cased workarounds from https://github.com/apple/pkl-intellij/pull/202[pkl-intellij] and https://github.com/apple/pkl-lsp/pull/188[pkl-lsp].

> * `pkl.ref#Domain.renderReference()` now accepts a `Reference<this, Any>` argument instead of `Reference<Domain, Any>`

Reviewer: Also `Any.ifNonNull()`:

```pkl
external function ifNonNull<Result>(transform: (this) -> Result): Result?
```

Not perfect; but much better than using `NonNull`

Maybe in the future we can introduce a type param for `NonNull` so this can become `NonNull<this>`
